src/solc_0.7/GameToken/GameToken.sol

```diff
@@ -0,0 +1,534 @@
+        }
+        require(assetIds.length > 0, "WITHDRAWAL_COMPLETE");
+        uint256[] memory values;
+        values = new uint256[](assetIds.length);
+        for (uint256 i = 0; i < assetIds.length; i++) {
+            values[i] = _gameAssets[baseId][assetIds[i]];
+            delete _gameAssets[baseId][assetIds[i]];
+        }
+        _asset.safeBatchTransferFrom(address(this), to, assetIds, values, "");
+
+        Update memory recovery;
+        recovery.assetIdsToRemove = assetIds;
+        recovery.assetAmountsToRemove = values;
+        emit GameTokenUpdated(gameId, 0, recovery, to);
+    }
+
+    /// @dev Check if a withdrawal is allowed.
+    /// @param from The address requesting the withdrawal.
+    /// @param gameId The id of the GAME token to withdraw assets from.
+    function _check_withdrawal_authorized(address from, uint256 gameId) internal view {
+        require(from != address(0), "SENDER_ZERO_ADDRESS");
+        require(from == _withdrawalOwnerOf(gameId), "LAST_OWNER_NOT_EQUAL_SENDER");
+    }
+
+    /// @dev Get the address allowed to withdraw assets from the GAME token.
+    /// If too many assets in GAME, block.gaslimit won't allow detroy and withdraw in 1 tx.
+    /// A game owner may destroy their GAME token, then withdraw assets in a later tx (even
+    /// though ownerOf(id) would be address(0) after burning.)
+    /// @param id The id of the GAME token to query.
+    /// @return the address of the owner before burning.
+    function _withdrawalOwnerOf(uint256 id) internal view returns (address) {
+        uint256 data = _owners[_storageId(id)];
+        if ((data & (2**160)) == 2**160) {
+            return address(data);
+        }
+        return address(0);
+    }
+
+    /// @dev Create a new gameId and associate it with an owner.
+    /// @param from The address of one creating the game.
+    /// @param to The address of the Game owner.
+    /// @param subId The id to use when generating the new GameId.
+    /// @param version The version number part of the gameId.
+    /// @param isCreation Whether this is a brand new GAME (as opposed to an update).
+    /// @return id The newly created gameId.
+    function _mintGame(
+        address from,
+        address to,
+        uint64 subId,
+        uint32 version,
+        bool isCreation
+    ) internal returns (uint256 id, uint256 baseId) {
+        uint32 idVersion;
+        uint256 gameId;
+        uint256 storageId;
+        uint256 data;
+        if (isCreation) {
+            idVersion = 1;
+            gameId = _generateGameId(from, subId, idVersion);
+            storageId = _storageId(gameId);
+            data = _owners[storageId];
+            require(data == 0, "BASEID_REUSE_FORBIDDEN");
+        } else {
+            idVersion = version;
+            gameId = _generateGameId(from, subId, idVersion);
+            storageId = _storageId(gameId);
+            data = _owners[storageId];
```

Comment:
data is not required here. we could also remove the local variavle data in the `isCreation` block and simply do `require(_owners[storageId] == 0, "BASEID_REUSE_FORBIDDEN");`